# Dataset Training using **Transformer**, **MLPMixer** and **HyperMixer**

### Importing Packages

In [1]:
import numpy as np
import pandas as pd
import random
import time
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from typing import Optional, Tuple

### Sentences and their Sentiment Labels lists

In [2]:
# Sentences
sentences = [
    "The sun is shining brightly today.",
    "I can't stand the constant noise from construction outside.",
    "This cake tastes delicious.",
    "My internet connection keeps dropping, it's so frustrating.",
    "The movie I watched last night was amazing.",
    "I'm feeling indifferent about what to have for dinner tonight.",
    "I got a promotion at work, I'm thrilled!",
    "The customer service representative was rude and unhelpful.",
    "Waking up to the sound of rain is so calming.",
    "I'm feeling under the weather today.",
    "The new restaurant in town has excellent food.",
    "The traffic this morning was terrible.",
    "Spending time with family always brings me joy.",
    "My computer crashed and I lost all my work, I'm devastated.",
    "I love the feeling of accomplishment after finishing a good book.",
    "The constant barking of my neighbor's dog is irritating.",
    "Traveling to new places is always exciting.",
    "I'm feeling indifferent about attending the party tonight.",
    "My favorite team won the championship, I'm ecstatic!",
    "The food at the party was disappointing.",
    "Watching the sunset by the beach is so peaceful.",
    "I had a terrible headache all day.",
    "Receiving unexpected gifts always puts a smile on my face.",
    "The rainy weather ruined my plans for a picnic.",
    "I love spending time with my pets, they bring me so much joy.",
    "Dealing with customer complaints all day is exhausting.",
    "The smell of fresh flowers is delightful.",
    "My car broke down on the way to work, what a disaster.",
    "Completing tasks on my to-do list gives me a sense of accomplishment.",
    "I'm feeling indifferent about which movie to watch tonight.",
    "I received a promotion at work, I'm over the moon!",
    "The food at the new restaurant was mediocre.",
    "Spending time in nature always rejuvenates me.",
    "I had a horrible experience at the dentist today.",
    "Watching my favorite TV show always cheers me up.",
    "The long queues at the grocery store are frustrating.",
    "Getting a good night's sleep makes me feel refreshed.",
    "I'm feeling indifferent about what to wear to the party.",
    "I passed my exam with flying colors, I'm elated!",
    "The service at the hotel was terrible.",
    "The smell of freshly baked cookies is heavenly.",
    "I'm feeling down because I missed my flight.",
    "Spending time with loved ones is priceless.",
    "My phone screen cracked, I'm so upset.",
    "Hearing my favorite song on the radio always lifts my spirits.",
    "The constant honking of horns in traffic is annoying.",
    "Enjoying a cup of hot chocolate on a cold day is comforting.",
    "I'm feeling indifferent about where to go for vacation.",
    "I received a promotion at work, I couldn't be happier!",
    "The food at the restaurant was subpar.",
    "Seeing a rainbow after the rain is magical.",
    "I had a terrible day at work.",
    "Receiving compliments from others always boosts my confidence.",
    "The loud music from the neighbor's party kept me awake all night.",
    "Reaching a personal goal brings a sense of fulfillment.",
    "I'm feeling indifferent about what book to read next.",
    "I won a prize in a contest, I'm thrilled!",
    "The customer service hotline was unresponsive.",
    "Watching a sunrise is a beautiful experience.",
    "I'm feeling upset because I missed my train.",
    "Achieving a personal best in a workout is empowering.",
    "I'm feeling indifferent about what to do this weekend.",
    "I received a compliment on my presentation, it made my day!",
    "The constant beeping of car horns in traffic is annoying.",
    "Visiting a botanical garden is a serene experience.",
    "I'm feeling frustrated because my internet is so slow.",
    "Enjoying a bubble bath after a long day is so relaxing.",
    "I'm feeling indifferent about which restaurant to choose for dinner.",
    "I won tickets to my favorite band's concert, I'm over the moon!",
    "The food at the party was tasteless and cold.",
    "Stargazing on a clear night is awe-inspiring.",
    "I'm feeling down because I lost my wallet.",
    "Receiving a handwritten letter from a friend brightened my day.",
    "The loud music from the neighbor's party is keeping me awake.",
    "Accomplishing a difficult task feels incredibly rewarding.",
    "I'm feeling indifferent about what to do this evening.",
    "I received praise from my boss for a job well done, I'm thrilled!",
    "The service at the restaurant was slow and the food was cold.",
    "The aroma of freshly brewed coffee is invigorating.",
    "I'm feeling frustrated because I can't find my keys.",
    "Taking a walk in the park on a sunny day is refreshing.",
    "I'm disappointed because my favorite show got canceled.",
    "Receiving a compliment from a stranger made my day.",
    "The traffic jam made me late for my appointment.",
    "I'm feeling indifferent about trying a new restaurant.",
    "The food at the party was tasteless and uninspiring.",
    "Listening to my favorite music always improves my mood.",
    "I'm feeling down because I missed the train.",
    "The beauty of nature never fails to amaze me.",
    "I'm upset because my plans got canceled at the last minute.",
    "Catching up with an old friend always brings joy.",
    "The constant beeping of car horns in traffic is irritating.",
    "Reading a good book transports me to another world.",
    "I'm feeling frustrated because of the long wait in line.",
    "The taste of homemade cookies is unparalleled.",
    "I'm excited because I'm going on vacation next week.",
    "Witnessing acts of kindness restores my faith in humanity.",
    "I'm feeling indifferent about what to do this weekend.",
    "The news of the unexpected promotion filled me with joy.",
    "The food at the restaurant was disappointing and overpriced.",
    "The flowers in the garden are blooming beautifully.",
    "I'm frustrated with the slow internet speed.",
    "Eating ice cream on a hot day is so refreshing.",
    "The constant noise from the construction site is unbearable.",
    "Winning the lottery would be a dream come true.",
    "I'm feeling indifferent about what to watch on TV tonight.",
    "Attending my best friend's wedding fills me with joy.",
    "The rude behavior of the cashier ruined my shopping experience.",
    "Listening to my favorite song always lifts my spirits.",
    "The long wait at the doctor's office is frustrating.",
    "Walking in the rain without an umbrella is not enjoyable.",
    "Receiving a compliment from a stranger brightens my day.",
    "I'm ecstatic about the upcoming vacation.",
    "The taste of burnt food is unpleasant.",
    "The breathtaking view from the mountaintop is awe-inspiring.",
    "Dealing with a broken phone is inconvenient.",
    "Playing with puppies brings me happiness.",
    "The constant honking of horns in traffic is annoying.",
    "Catching up with old friends is always fun.",
    "The unexpected rain ruined our picnic plans.",
    "Getting a promotion at work is exciting.",
    "The smell of fresh bread baking is delightful.",
    "I'm feeling down because I lost my wallet.",
    "Watching a sunset on the beach is peaceful.",
    "The loud music from the party next door is disturbing.",
    "Reading a good book before bed helps me relax.",
    "I'm frustrated with the lack of communication from my partner.",
    "Seeing a shooting star is a rare and magical experience.",
    "The broken elevator in the building is inconvenient.",
    "Receiving a handwritten letter from a friend warms my heart.",
    "I'm overjoyed to be spending time with my family.",
    "The taste of bitter medicine is unpleasant.",
    "The beautiful colors of autumn leaves are mesmerizing.",
    "Dealing with rude customers is exhausting.",
    "Finding money in an old jacket pocket is a pleasant surprise.",
    "I'm feeling indifferent about going to the gym today.",
    "The smell of fresh coffee in the morning is invigorating.",
    "I'm disappointed because my favorite restaurant is closed.",
    "Spending time in nature helps me relax and recharge.",
    "The constant ringing of the phone is irritating.",
    "Receiving a hug from a loved one is comforting.",
    "I'm excited to try the new restaurant in town.",
    "The taste of homemade cookies is unbeatable.",
    "I'm upset because my flight got canceled.",
    "Watching a comedy show always makes me laugh.",
    "The never-ending emails in my inbox are overwhelming.",
    "Receiving a thoughtful gift from a friend is heartwarming.",
    "I'm indifferent about what to cook for dinner tonight."

]

# Sentiment Labels
sentiment_labels = [
    "Positive", "Negative", "Positive", "Negative", "Positive",
    "Neutral", "Positive", "Negative", "Positive", "Negative",
    "Positive", "Negative", "Positive", "Negative", "Positive",
    "Negative", "Positive", "Neutral", "Positive", "Negative",
    "Positive", "Negative", "Positive", "Negative", "Positive",
    "Negative", "Positive", "Negative", "Positive", "Neutral",
    "Positive", "Negative", "Positive", "Negative", "Positive",
    "Negative", "Positive", "Neutral", "Positive", "Negative",
    "Positive", "Negative", "Positive", "Negative", "Positive",
    "Negative", "Positive", "Negative", "Positive", "Neutral",
    "Positive", "Negative", "Positive", "Negative", "Positive",
    "Negative", "Positive", "Negative", "Positive", "Neutral",
    "Positive", "Negative", "Positive", "Negative", "Positive",
    "Negative", "Positive", "Negative", "Positive", "Neutral",
    "Positive", "Negative", "Positive", "Negative", "Positive",
    "Negative", "Positive", "Negative","Positive", "Negative",
    "Positive", "Negative", "Positive","Negative", "Neutral",
    "Negative", "Positive", "Negative","Positive", "Negative",
    "Positive", "Negative", "Positive","Negative", "Positive",
    "Positive", "Positive", "Neutral","Positive", "Negative",
    "Positive", "Negative", "Positive", "Negative", "Positive",
    "Neutral", "Positive", "Negative", "Positive", "Negative",
    "Negative", "Positive", "Positive", "Negative", "Positive",
    "Negative", "Positive", "Negative", "Positive", "Negative",
    "Positive", "Positive", "Negative", "Positive", "Negative",
    "Negative", "Positive", "Positive", "Negative", "Positive",
    "Positive", "Negative", "Positive", "Negative", "Negative",
    "Positive", "Positive", "Negative", "Positive", "Positive",
    "Negative", "Positive", "Negative", "Positive", "Negative",
    "Positive", "Positive", "Negative",
]

len(sentences), len(sentiment_labels)

(148, 148)

In [3]:
# Count occurrences of each label
positive_count = sentiment_labels.count("Positive")
negative_count = sentiment_labels.count("Negative")
neutral_count = sentiment_labels.count("Neutral")

# Print the counts
positive_count, negative_count, neutral_count

(76, 62, 10)


### 1.  Creating a dataset of 148 sentences with sentiment labels as either **positive, negative** or **neutral**.
### 2.    Then creating a training and test set respectively


In [4]:
# Create DataFrame
df = pd.DataFrame({'Sentence': sentences, 'Sentiment': sentiment_labels})

# Shuffle the Dataset
random.shuffle(df.values)

# Split the dataset into training and test sets
train_df, test_df = train_test_split(df, test_size=0.2, random_state=42)

# Print training set
print(f"Training Set Size:{train_df.shape}")

# Print test set
print(f"Test Set Size:{test_df.shape}")

Training Set Size:(118, 2)
Test Set Size:(30, 2)


In [5]:
MLP_mixer_accuracies=[]
MLP_mixer_runtimes = []

Transformer_accuracies=[]
Transformer_runtimes = []

HyperMixer_accuracies=[]
HyperMixer_runtimes = []

n=50

### MLP Mixer

In [6]:
# Define MLP-Mixer model
class MLP_Mixer(nn.Module):
    def __init__(self, input_dim, hidden_dim, num_classes):
        super(MLP_Mixer, self).__init__()
        self.fc1 = nn.Linear(input_dim, hidden_dim)
        self.fc2 = nn.Linear(hidden_dim, hidden_dim)
        self.fc3 = nn.Linear(hidden_dim, num_classes)

    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        x = self.fc3(x)
        return x

# Encode labels
label_encoder = LabelEncoder()
train_df['Label'] = label_encoder.fit_transform(train_df['Sentiment'])

# Convert sentences to numerical representation
vocab = set(" ".join(train_df['Sentence']).split())
word_to_index = {word: i for i, word in enumerate(vocab)}
def sentence_to_one_hot(sentence):
    one_hot = [0] * len(vocab)
    for word in sentence.split():
        one_hot[word_to_index[word]] = 1
    return one_hot

train_data = torch.tensor([sentence_to_one_hot(sentence) for sentence in train_df['Sentence']], dtype=torch.float32)
train_labels = torch.tensor(train_df['Label'].values, dtype=torch.long)

# Define hyperparameters
input_dim = len(vocab)  # Vocabulary size
hidden_dim = 128
num_classes = len(label_encoder.classes_)  # Number of unique classes

for i  in range(n):
    # Instantiate MLP-Mixer model
    model = MLP_Mixer(input_dim, hidden_dim, num_classes)

    # Define loss function and optimizer
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=0.001)

    accuracies_MLPMixer=[]
    start_time = time.time()  # Start time for each run

    # Train the model
    num_epochs = 5
    for epoch in range(num_epochs):
        optimizer.zero_grad()
        outputs = model(train_data)
        loss = criterion(outputs, train_labels)
        loss.backward()
        optimizer.step()

        # Calculate accuracy
        _, predicted = torch.max(outputs, 1)
        accuracy = accuracy_score(train_labels.numpy(), predicted.numpy())
        accuracies_MLPMixer.append(accuracy)
        print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}, Accuracy: {accuracy:.4f}')

    end_time = time.time()  # End time for each run
    runtime = end_time - start_time
    MLP_mixer_accuracies.append(np.mean(accuracies_MLPMixer))
    MLP_mixer_runtimes.append(runtime)
    print(f"{i+1} Run Complete!\n")

Epoch [1/5], Loss: 1.1268, Accuracy: 0.0593
Epoch [2/5], Loss: 1.1126, Accuracy: 0.2797
Epoch [3/5], Loss: 1.0992, Accuracy: 0.3898
Epoch [4/5], Loss: 1.0860, Accuracy: 0.3814
Epoch [5/5], Loss: 1.0728, Accuracy: 0.4237
1 Run Complete!

Epoch [1/5], Loss: 1.1319, Accuracy: 0.0508
Epoch [2/5], Loss: 1.1149, Accuracy: 0.1186
Epoch [3/5], Loss: 1.0989, Accuracy: 0.3475
Epoch [4/5], Loss: 1.0834, Accuracy: 0.4322
Epoch [5/5], Loss: 1.0682, Accuracy: 0.6441
2 Run Complete!

Epoch [1/5], Loss: 1.0742, Accuracy: 0.5678
Epoch [2/5], Loss: 1.0597, Accuracy: 0.5678
Epoch [3/5], Loss: 1.0461, Accuracy: 0.5678
Epoch [4/5], Loss: 1.0329, Accuracy: 0.5678
Epoch [5/5], Loss: 1.0198, Accuracy: 0.5678
3 Run Complete!

Epoch [1/5], Loss: 1.1030, Accuracy: 0.5254
Epoch [2/5], Loss: 1.0871, Accuracy: 0.5678
Epoch [3/5], Loss: 1.0721, Accuracy: 0.5678
Epoch [4/5], Loss: 1.0574, Accuracy: 0.5678
Epoch [5/5], Loss: 1.0427, Accuracy: 0.5678
4 Run Complete!

Epoch [1/5], Loss: 1.0774, Accuracy: 0.3814
Epoch [2

### Transformers

In [7]:
# Define Transformer model
class TransformerModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, num_classes, num_layers=1, num_heads=1, dropout=0.1):
        super(TransformerModel, self).__init__()
        self.embedding = nn.Embedding(input_dim, hidden_dim)
        self.encoder_layer = nn.TransformerEncoderLayer(hidden_dim, num_heads)
        self.transformer_encoder = nn.TransformerEncoder(self.encoder_layer, num_layers)
        self.fc = nn.Linear(hidden_dim, num_classes)

    def forward(self, x):
        x = self.embedding(x)
        x = x.permute(1, 0, 2)  # Shape: (seq_len, batch_size, embedding_dim)
        output = self.transformer_encoder(x)
        output = output.mean(dim=0)  # Average over sequence length
        output = self.fc(output)
        return output

# Encode labels
label_encoder = LabelEncoder()
train_df['Label'] = label_encoder.fit_transform(train_df['Sentiment'])

# Convert sentences to numerical representation
vocab = set(" ".join(train_df['Sentence']).split())
word_to_index = {word: i for i, word in enumerate(vocab)}
def sentence_to_indices(sentence):
    indices = [word_to_index[word] for word in sentence.split()]
    return indices

train_data_indices = [sentence_to_indices(sentence) for sentence in train_df['Sentence']]

# Pad sequences to the same length
max_seq_length = max(len(indices) for indices in train_data_indices)
train_data_padded = [indices + [0] * (max_seq_length - len(indices)) for indices in train_data_indices]

# Convert data and labels to PyTorch tensors
train_data_tensor = torch.tensor(train_data_padded, dtype=torch.long)
train_labels_tensor = torch.tensor(train_df['Label'].values, dtype=torch.long)

# Define hyperparameters
input_dim = len(vocab)  # Vocabulary size
hidden_dim = 128
num_classes = len(label_encoder.classes_)  # Number of unique classes
num_layers = 2
num_heads = 4
dropout = 0.1

for i in range(n):
    # Instantiate Transformer model
    model = TransformerModel(input_dim, hidden_dim, num_classes, num_layers, num_heads, dropout)

    # Define loss function and optimizer
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=0.001)

    # Prepare DataLoader
    batch_size = 2
    train_dataset = TensorDataset(train_data_tensor, train_labels_tensor)
    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)

    accuracies_Transformer=[]
    start_time = time.time()  # Start time for each run

    # Train the model
    num_epochs = 5
    for epoch in range(num_epochs):
        model.train()
        total_loss = 0
        total_correct = 0
        total_samples = 0
        for inputs, targets in train_loader:
            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            loss.backward()
            optimizer.step()

            total_loss += loss.item() * len(inputs)
            _, predicted = torch.max(outputs, 1)
            total_correct += (predicted == targets).sum().item()
            total_samples += len(inputs)

        epoch_loss = total_loss / total_samples
        epoch_accuracy = total_correct / total_samples
        accuracies_Transformer.append(epoch_accuracy)
        print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {epoch_loss:.4f}, Accuracy: {epoch_accuracy:.4f}')

    end_time = time.time()  # End time for each run
    runtime = end_time - start_time
    Transformer_runtimes.append(runtime)
    Transformer_accuracies.append(np.mean(accuracies_Transformer))
    print(f"{i+1} Run Complete!\n")

/usr/local/lib/python3.10/dist-packages/torch/nn/modules/transformer.py:286: UserWarning: enable_nested_tensor is True, but self.use_nested_tensor is False because encoder_layer.self_attn.batch_first was not True(use batch_first for better inference performance)
  warnings.warn(f"enable_nested_tensor is True, but self.use_nested_tensor is False because {why_not_sparsity_fast_path}")


Epoch [1/5], Loss: 1.0659, Accuracy: 0.4915
Epoch [2/5], Loss: 0.6736, Accuracy: 0.7034
Epoch [3/5], Loss: 0.3680, Accuracy: 0.8814
Epoch [4/5], Loss: 0.2252, Accuracy: 0.9322
Epoch [5/5], Loss: 0.1603, Accuracy: 0.9492
1 Run Complete!

Epoch [1/5], Loss: 0.9711, Accuracy: 0.5000
Epoch [2/5], Loss: 0.9399, Accuracy: 0.5169
Epoch [3/5], Loss: 0.5739, Accuracy: 0.8220
Epoch [4/5], Loss: 0.3655, Accuracy: 0.9068
Epoch [5/5], Loss: 0.2060, Accuracy: 0.9322
2 Run Complete!

Epoch [1/5], Loss: 1.0301, Accuracy: 0.4661
Epoch [2/5], Loss: 0.7284, Accuracy: 0.6356
Epoch [3/5], Loss: 0.3530, Accuracy: 0.8898
Epoch [4/5], Loss: 0.2499, Accuracy: 0.9153
Epoch [5/5], Loss: 0.2459, Accuracy: 0.9237
3 Run Complete!

Epoch [1/5], Loss: 0.9966, Accuracy: 0.4831
Epoch [2/5], Loss: 0.5112, Accuracy: 0.7966
Epoch [3/5], Loss: 0.2602, Accuracy: 0.9068
Epoch [4/5], Loss: 0.1367, Accuracy: 0.9576
Epoch [5/5], Loss: 0.1877, Accuracy: 0.9322
4 Run Complete!

Epoch [1/5], Loss: 1.0470, Accuracy: 0.5424
Epoch [2

### HyperMixer

In [8]:
# Define HyperMixer model
class HyperMixerModel(nn.Module):
    def __init__(
            self,
            input_dim: int,
            hidden_dim: int,
            num_classes: int,
            norm_layer: type = nn.LayerNorm,
            act_layer: type = nn.GELU,
            drop: float = 0.
    ) -> None:
        super(HyperMixerModel, self).__init__()
        self.embedding = nn.Embedding(input_dim, hidden_dim)
        self.mixer_block = HyperMixerBlock(hidden_dim, norm_layer=norm_layer, act_layer=act_layer, drop=drop)
        self.fc = nn.Linear(hidden_dim, num_classes)

    def forward(self, x):
        x = self.embedding(x)
        x = self.mixer_block(x)
        x = torch.mean(x, dim=1)  # Average over tokens
        x = self.fc(x)
        return x

# Define HyperMixerBlock
class HyperMixerBlock(nn.Module):
    def __init__(
            self,
            hidden_dim: int,
            norm_layer: type = nn.LayerNorm,
            act_layer: type = nn.GELU,
            drop: float = 0.
    ) -> None:
        super(HyperMixerBlock, self).__init__()
        self.norm = norm_layer(hidden_dim)
        self.mlp_tokens = nn.Sequential(
            nn.Linear(hidden_dim, hidden_dim * 2),
            act_layer(),
            nn.Dropout(drop),
            nn.Linear(hidden_dim * 2, hidden_dim),
            nn.Dropout(drop)
        )
        self.mlp_channels = nn.Sequential(
            nn.Linear(hidden_dim, hidden_dim * 2),
            act_layer(),
            nn.Dropout(drop),
            nn.Linear(hidden_dim * 2, hidden_dim),
            nn.Dropout(drop)
        )

    def forward(
            self,
            x: torch.Tensor
    ) -> torch.Tensor:
        x = self.norm(x)
        x = x + self.mlp_tokens(x)
        x = x + self.mlp_channels(x.transpose(1, 2)).transpose(1, 2)
        return x

# Encode labels
label_encoder = LabelEncoder()
train_df['Label'] = label_encoder.fit_transform(train_df['Sentiment'])

# Convert sentences to numerical representation
vocab = set(" ".join(train_df['Sentence']).split())
word_to_index = {word: i for i, word in enumerate(vocab)}
def sentence_to_indices(sentence):
    indices = [word_to_index[word] for word in sentence.split()]
    return indices

train_data_indices = [sentence_to_indices(sentence) for sentence in train_df['Sentence']]

# Pad sequences to the same length
max_seq_length = max(len(indices) for indices in train_data_indices)
train_data_padded = [indices + [0] * (max_seq_length - len(indices)) for indices in train_data_indices]

# Convert data and labels to PyTorch tensors
train_data_tensor = torch.tensor(train_data_padded, dtype=torch.long)
train_labels_tensor = torch.tensor(train_df['Label'].values, dtype=torch.long)

# Define hyperparameters
input_dim = len(vocab)  # Vocabulary size
hidden_dim = 13
num_classes = len(label_encoder.classes_)  # Number of unique classes

for i in range(n):
    # Instantiate HyperMixer model
    model = HyperMixerModel(input_dim, hidden_dim, num_classes)

    # Define loss function and optimizer
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=0.001)

    # Prepare DataLoader
    batch_size = 2  # Adjusted for small dataset
    train_dataset = TensorDataset(train_data_tensor, train_labels_tensor)
    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)

    accuracies_HyperMixer=[]
    start_time = time.time()  # Start time for each run

    # Train the model
    num_epochs = 5
    for epoch in range(num_epochs):
        model.train()
        total_loss = 0
        total_correct = 0
        total_samples = 0
        for inputs, targets in train_loader:
            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            loss.backward()
            optimizer.step()

            total_loss += loss.item() * len(inputs)
            _, predicted = torch.max(outputs, 1)
            total_correct += (predicted == targets).sum().item()
            total_samples += len(inputs)

        epoch_loss = total_loss / total_samples
        epoch_accuracy = total_correct / total_samples
        accuracies_HyperMixer.append(epoch_accuracy)
        print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {epoch_loss:.4f}, Accuracy: {epoch_accuracy:.4f}')

    end_time = time.time()  # End time for each run
    runtime = end_time - start_time
    HyperMixer_runtimes.append(runtime)
    HyperMixer_accuracies.append(np.mean(accuracies_HyperMixer))
    print(f"{i+1} Run Complete!\n")

Epoch [1/5], Loss: 1.0746, Accuracy: 0.3814
Epoch [2/5], Loss: 0.9130, Accuracy: 0.5763
Epoch [3/5], Loss: 0.8342, Accuracy: 0.5593
Epoch [4/5], Loss: 0.7724, Accuracy: 0.6610
Epoch [5/5], Loss: 0.6740, Accuracy: 0.7542
1 Run Complete!

Epoch [1/5], Loss: 1.0505, Accuracy: 0.5169
Epoch [2/5], Loss: 0.9056, Accuracy: 0.5678
Epoch [3/5], Loss: 0.7828, Accuracy: 0.6271
Epoch [4/5], Loss: 0.7014, Accuracy: 0.7712
Epoch [5/5], Loss: 0.5787, Accuracy: 0.7966
2 Run Complete!

Epoch [1/5], Loss: 0.9183, Accuracy: 0.5763
Epoch [2/5], Loss: 0.8423, Accuracy: 0.5763
Epoch [3/5], Loss: 0.7940, Accuracy: 0.6186
Epoch [4/5], Loss: 0.7246, Accuracy: 0.6610
Epoch [5/5], Loss: 0.5754, Accuracy: 0.8136
3 Run Complete!

Epoch [1/5], Loss: 0.9752, Accuracy: 0.5508
Epoch [2/5], Loss: 0.8486, Accuracy: 0.5763
Epoch [3/5], Loss: 0.7752, Accuracy: 0.6017
Epoch [4/5], Loss: 0.6946, Accuracy: 0.7203
Epoch [5/5], Loss: 0.5514, Accuracy: 0.8390
4 Run Complete!

Epoch [1/5], Loss: 0.9262, Accuracy: 0.5763
Epoch [2

In [9]:
import plotly.graph_objs as go

number_of_runs = list(range(1, n+1))

# Assuming you have defined HyperMixer_accuracies, Transformer_accuracies, and MLP_mixer_accuracies somewhere

# Plotting
trace1 = go.Scatter(x=number_of_runs, y=HyperMixer_accuracies, mode='lines+markers', name='HyperMixer', line=dict(width=5, dash='solid'))
trace2 = go.Scatter(x=number_of_runs, y=Transformer_accuracies, mode='lines+markers', name='Transformers', line=dict(color='rgb(250, 100, 147)', width=2, dash='solid'))
trace3 = go.Scatter(x=number_of_runs, y=MLP_mixer_accuracies, mode='lines+markers', name='MLPMixer')


layout = go.Layout(
    title='Model Accuracies',
    xaxis=dict(title='Number of Runs'),
    yaxis=dict(title='Accuracy'),
    legend=dict(x=0, y=1.0),
    hovermode='closest',
    plot_bgcolor='rgba(0,0,0,0)',
)

fig = go.Figure(data=[trace1, trace2, trace3], layout=layout)

# To display the plot in Jupyter Notebook or Jupyter Lab
# If you're using a different environment, you may need to use the appropriate display function.
fig.show()

In [10]:
#Average Accuracies

np.mean(MLP_mixer_accuracies), np.mean(Transformer_accuracies),np.mean(HyperMixer_accuracies)

(0.5131525423728814, 0.768, 0.6302372881355932)

In [11]:
import plotly.graph_objs as go

number_of_runs = list(range(1, n+1))

# Assuming you have defined MLP_mixer_runtimes, Transformer_runtimes, and HyperMixer_runtimes somewhere

# Plotting
trace1 = go.Scatter(x=number_of_runs, y=HyperMixer_runtimes, mode='lines+markers', name='HyperMixer', line=dict(width=5, dash='solid'))
trace2 = go.Scatter(x=number_of_runs, y=Transformer_runtimes, mode='lines+markers', name='Transformers', line=dict(color='rgb(250, 100, 147)', width=2, dash='solid'))
trace3 = go.Scatter(x=number_of_runs, y=MLP_mixer_runtimes, mode='lines+markers', name='MLPMixer')


layout = go.Layout(
    title='Model Runtimes',
    xaxis=dict(title='Number of Runs'),
    yaxis=dict(title='Runtime'),
    legend=dict(x=0, y=1.0),
    hovermode='closest',
    plot_bgcolor='rgba(0,0,0,0)',
)

fig = go.Figure(data=[trace1, trace2, trace3], layout=layout)

fig.show()

In [12]:
#Average Runtime

np.mean(MLP_mixer_runtimes), np.mean(Transformer_runtimes),np.mean(HyperMixer_runtimes)

(0.018990530967712402, 4.825806703567505, 0.8242124509811402)